kratos/containers/sparse_contiguous_row_graph.h

```diff
@@ -209,58 +210,84 @@ class SparseContiguousRowGraph
         return mGraph;
     }
 
+    //note that this function is slightly less efficient than the span version below
     template<class TVectorType=DenseVector<IndexType>>
     IndexType ExportCSRArrays(
         TVectorType& rRowIndices,
         TVectorType& rColIndices
     ) const
+    {
+        IndexType* pRowIndicesData=nullptr;
+        IndexType RowIndicesDataSize=0;
+        IndexType* pColIndicesData=nullptr;
+        IndexType ColIndicesDataSize=0;
+        ExportCSRArrays(pRowIndicesData,RowIndicesDataSize,pColIndicesData,ColIndicesDataSize);
+        rRowIndices = span<IndexType>(pRowIndicesData,RowIndicesDataSize); //here a copy happens!
+        rColIndices = span<IndexType>(pColIndicesData,ColIndicesDataSize);; //here a copy happens!
+        return rRowIndices.size();
+    }
+
+    IndexType ExportCSRArrays(
+        IndexType*& pRowIndicesData,
+        IndexType& rRowDataSize,
+        IndexType*& pColIndicesData,
+        IndexType& rColDataSize
+    ) const
     {
         //need to detect the number of rows this way since there may be gaps
         IndexType nrows=Size();
 
-        if(rRowIndices.size() != nrows+1)
+        pRowIndicesData = new IndexType[nrows+1];
+        rRowDataSize=nrows+1;
+        span<IndexType> row_indices(pRowIndicesData, nrows+1);
+
+        if(nrows == 0) //empty
         {
-            rRowIndices.resize(nrows+1, false);
+            row_indices[0] = 0;
         }
-        //set it to zero in parallel to allow first touching
-        IndexPartition<IndexType>(nrows+1).for_each([&](IndexType i){
-            rRowIndices[i] = 0;
-        });
-
-        //count the entries 
-        IndexPartition<IndexType>(nrows).for_each([&](IndexType i){
-            rRowIndices[i+1] = mGraph[i].size();
-        });
-        
-        //sum entries
-        for(IndexType i = 1; i<static_cast<IndexType>(rRowIndices.size()); ++i){
-            rRowIndices[i] += rRowIndices[i-1];
+        else
+        {
+            //set it to zero in parallel to allow first touching
+            IndexPartition<IndexType>(nrows+1).for_each([&](IndexType i){
+                row_indices[i] = 0;
+            });
+
+            //count the entries 
+            IndexPartition<IndexType>(nrows).for_each([&](IndexType i){
+                row_indices[i+1] = mGraph[i].size();
+            });
+            
+            //sum entries
+            for(IndexType i = 1; i<static_cast<IndexType>(row_indices.size()); ++i){
+                row_indices[i] += row_indices[i-1];
+            }
         }
 
-        IndexType nnz = rRowIndices[nrows];
-        if(rColIndices.size() != nnz){
-            rColIndices.resize(nnz, false);
-        }
+        IndexType nnz = row_indices[nrows];
+        rColDataSize=nnz;
+        pColIndicesData = new IndexType[nnz];
```

Comment:
same

Reply:
the ownership is transferred to the caller, who is in charge of deleting those arrays...